### DISCUSSION TITLE
NX support

### DISCUSSION THREAD

**Discussion by debrouxl** [memtest86plus/memtest86plus#241] [Enhancement]
NX support was mentioned in #34 as future work and a possible future requirement for Secure Boot ("will eventually need to support being loaded with NX (W^X) vulnerability mitigations enabled"), but postponed.

As a learning exercise, I have implemented NX for pd2 (unconditional right from the startup code on x86_64, and only from `map_window()` on x86), and pushed the result on the "nx" branch ( https://github.com/memtest86plus/memtest86plus/tree/nx ).
For now, I took the conservative route with separate long mode enablement and NX enablement, but FWIW, the only real CPU on which I tested simultaneous enablement, a dual-core K8 (so both the BSP and AP code paths were taken), didn't crash.

**Comment by x86fr** [comment]
NX Bit was implemented as a mitigation to avoid security exploit (by executing malicious code on a memory area normally reserved for datas).  Why would we need NX support for a tool like Memtest86+, especially after UEFI `ExitBootServices()`? I don't get it.

**Comment by debrouxl** [comment]
Well, I've just found https://github.com/rhboot/shim-review/issues/307 : NX support has become a requirement for getting a binary signed with the MS key. So I think that this is what was referred to in #34.

I'm not stating that _technically_, full NX support on all but the memory area from which the Memtest86+ code is currently running + other areas it depends on, makes Memtest86+ really much safer in practice, since it's not a general-purpose OS accepting arbitrary input from arbitrary user stimuli on many classes of hardware. After `ExitBootServices()`, our TCB becomes quite small :)
However, being signed with the MS key would boost practical usability of the memtest86+ binary releases. Put another way: not being signable and eventually signed with the MS key will hamper out of the box usability by end users, and may make the integration of memtest86+ into some software distributions more difficult or impossible.
And implementing partial NX support on pd2, i.e. the largest part of the memory on computers with >= 6 GB of RAM with the current memory mapping, wasn't technically hard. My patch might even be overly complex, if x86_64 CPUs support LM and NX being enabled simultaneously by a single MSR write.

**Comment by x86fr** [comment]
NX Bit must be set on the **shim** build when submitted to Microsoft for signing. I don't think it is required for all the software chainloaded after shim, but if we need to add the flag on our PE binary, it's probably not very complex. Things are different if we want our binary to be signed directly by Microsoft, but I don't see the point of doing that. Signing by Microsoft is a lengthy process and using shim seems much easier to maintain (as we will be able to release a signed binary at any time without delay).

**Comment by martinwhitaker** [comment]
As I read it, all we need to do is make sure our code runs when loaded with NX protection enabled, up until we rewrite the page tables. If loaded via shim, that's the code starting from `efi_handover` in `startup64.S` and ending when we load the GDT. That includes the call to `efi_setup`. If loaded directly from the BIOS, that also includes the four extra instructions starting at `efi_boot`.

I can't see any real benefit from enabling NX protection once we've started up.
